lib/eco/reducer.py

```diff
@@ -0,0 +1,153 @@
+            if self.test == self.original:
+                logging.info("No more single reductions possible.")
+                break
+
+    def reduce(self, mode):
+        self.pos = 0
+        logging.info("Pos: {} Len: {} Start {}-reduce".format(self.pos, len(self.test), mode))
+        try:
+            # Run without changes to store exception
+            self.run()
+        except Exception as e:
+            self.lastexception = e
+        while self.pos < len(self.test):
+            self.delete(mode)
+            try:
+                self.run() # check if first run causes error
+                logging.info("Pos: {} Len: {} Fixed".format(self.pos, len(self.test)))
+                self.revert()
+            except Exception as e:
+                if not self.same_exception(e):
+                    logging.info("Pos: {} Len: {} Wrong exception: '{}'".format(self.pos, len(self.test), e))
+                    self.revert()
+                    continue
+                else:
+                    logging.info("Pos: {} Len: {} Still failing: '{}'".format(self.pos, len(self.test), e))
+                    self.save() # save progress
+        self.save()
+
+    def same_exception(self, e):
+        if not self.lastexception:
+            self.lastexception = e
+            return True
+        if type(self.lastexception) is type(e) and \
+                self.lastexception.args == e.args:
+                    return True
+        return False
+
+    def delete(self, mode):
+        self.old = self.test[:] # copy
+        if mode == "single":
+            del self.test[self.pos:self.pos+4] # remove one keystroke
+            if self.pos < len(self.test) and \
+                    self.test[self.pos] == "self.treemanager.undo_snapshot()":
+                        self.test.pop(self.pos)
+        elif mode == "snapshot":
+            del self.test[self.pos:self.pos+81] # remove one snapshot
+
+    def revert(self):
+        diff = len(self.old) - len(self.test)
+        self.test = self.old[:] # revert changes
+        self.pos += diff # and skip test lines that are needed
+
+    def set_destination(self, dest):
+        filename = dest + ".1"
+        i = 1
+        while os.path.isfile(filename):
+            filename = dest + ".{}".format(i)
+            i += 1
+        self.dest = filename
+        logging.info("Setting destination to '{}'".format(filename))
+
+    def save(self):
+        f = open(self.dest, "w")
+        for l in self.test:
+            f.write(l)
+            f.write("\n")
+        f.close()
+
+    def run(self):
+        self.reset()
+        self.treemanager.import_file(self.program)
+        for l in self.test:
+            exec(l)
+
+    def move(self, direction, times):
+        for i in range(times): self.treemanager.cursor_movement(direction)
+
+optp = OptionParser(usage="usage: python2 %prog PROGRAM TEST LANG [options]\n\ne.g. python2 %prog program.txt tests.txt \"Python 2.7.5\"")
```

Comment:
Probably best to wrap this in a `if __name__ == "__main__"`, just in case you want to import this as a library one day.